File: 22-generate-parentheses/22-generate-parentheses.swift
class Solution {
    func backtrack(_ current: String, _ open: Int, _ close: Int, _ n: Int, _ result: (String) -> (Void)){
        if(current.count == 2 * n){
            result(current)
            return
        }
        if(open != n){
            backtrack(current + "(", open + 1, close, n, result)
        }
        if(open != close){
            backtrack(current + ")", open, close + 1, n, result)
        }
    }
    func generateParenthesis(_ n: Int) -> [String] {
        var result = [String]()
        backtrack("", 0, 0, n){result.append($0)}
        return result
    }
}

File: 22-generate-parentheses/22-generate-parentheses.cpp
class Solution {
public:
    void backtrack(vector<string>& ans,int open, int close, string curr, int n){
        if(curr.length() == 2*n){
            ans.push_back(curr);
            return;
        }
        if(open != n)backtrack(ans, open + 1, close, curr + "(", n);
        if(open != close)backtrack(ans, open, close + 1, curr + ")", n);
    }
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        backtrack(ans, 0, 0, "", n);
        return ans;
    }
};